File: AlgorithmsPackCPP/Sliding Puzzle.cpp
class Solution {
    private:
    string goal = "123450"; 
    vector<vector<int>>dir{{1,3},{0,2,4},{1,5},{0,4},{1,3,5},{2,4}};   
//This works in a similar way to flattening the 2D array.
//Index 0 has neighbours 1/3, 1 has neighbours 0/2/4, and so on.
public:
    int slidingPuzzle(vector<vector<int>>& board) {
        string start = "";  
        for(int i=0; i < board.size(); i++){
            for(int j = 0; j < board[0].size(); j++){
                start+=board[i][j]+'0'; 
            }
        }
        
        if(start==goal) 
            return 0;
        
        unordered_set<string>visited;
        queue<string>q;
        q.push(start);
        visited.insert(start);
        
        for(int level = 0, sz = q.size(); !q.empty(); ++level, sz = q.size()){
            while(sz--){
                string cur = q.front();
                q.pop();
                int idx1 = cur.find('0');
                for(int idx2:dir[idx1]){ 
                    swap(cur[idx1],cur[idx2]);  
                    if(cur==goal) 
                        return level+1;
                    
                    if(visited.insert(cur).second) 
                        q.push(cur); 
                    
                    swap(cur[idx1],cur[idx2]); 
                }
            }
        }
        return -1;
    }
};

File: AlgorithmsPackPythonSolutions/Sliding Puzzle.py
class Solution:
    def slidingPuzzle(self, board: List[List[int]]) -> int:
        goal = "123450"
        directions = [[1, 3], [0, 2, 4], [1, 5], [0, 4], [1, 3, 5], [2, 4]]
        start = ""
        
        for i in range(len(board)):
            for j in range(len(board[0])):
                start += str(board[i][j])
        
        if start == goal:
            return 0
        
        visited = set()
        q = deque()
        q.append(start)
        visited.add(start)
        
        level = 0
        while q:
            size = len(q)
            for _ in range(size):
                cur = q.popleft()
                idx1 = cur.index('0')
                
                for idx2 in directions[idx1]:
                    cur = list(cur)
                    cur[idx1], cur[idx2] = cur[idx2], cur[idx1]
                    cur_str = "".join(cur)
                    
                    if cur_str == goal:
                        return level + 1
                    
                    if cur_str not in visited:
                        visited.add(cur_str)
                        q.append(cur_str)
                    
                    cur[idx1], cur[idx2] = cur[idx2], cur[idx1] 
                    
            level += 1
        
        return -1
